### ISSUE TITLE
Computed Role is distinct from WAI-ARIA role

### ISSUE THREAD

**Issue by yezhizhen** [servo/servo#43734]
**Describe the bug:**
[Computed Role](https://www.w3.org/TR/core-aam-1.2/#roleMappingComputedRole) is intended for devtool and automated testing.

You can define [ARIA attribute](https://w3c.github.io/aria/#introroles) all you like. But computed role has its own rules. Right now we use ARIA attribute as our computed role.

The basic rule is actually not complex to implement, but a bit tedious:
You would need to refer to 
1. [HTML Accessibility API Mappings](https://www.w3.org/TR/html-aam-1.0/#html-element-role-mappings) for role mapping table.
2. If no fallback is found, use `generic` role.

See 3 examples [here](https://www.w3.org/TR/core-aam-1.2/#roleMappingComputedRole) for a better hang on this.

**To Reproduce:**
We are failing https://github.com/servo/servo/blob/bc9a86f7bd57f19b89d533b6b1ed1b62f95f2a69/tests/wpt/tests/webdriver/tests/classic/get_computed_role/get.py#L78-L82
for article.

The ARIA attribute is None. But computed role should be article here, according to https://www.w3.org/TR/html-aam-1.0/#el-article

cc @xiaochengh

**Comment by lolaodelola** [comment]
I'm going to investigate this and potentially take it on (see: [chat](https://servo.zulipchat.com/#narrow/channel/263398-general/topic/accessibility.20support/with/598122973))

**Comment by lolaodelola** [comment]
Hi folks, this is my rough proposal for this:

- In [webdriver_handler](https://github.com/servo/servo/blob/bc9a86f7bd57f19b89d533b6b1ed1b62f95f2a69/components/script/webdriver_handlers.rs#L1355) `handle_get_computed_role`
  - Check the explicit "role" of the element (currently happening)
  - If there is no role, use the implicit role as defined in HTML-AAM. The accessibility tree assigns the node's role using the [`HMTL_ROLE_MAPPINGS`](https://github.com/servo/servo/blob/c2a3dbc1384b1e2e155a653ca7d42cf3660fe0c6/components/layout/accessibility_tree.rs#L669) hashmap, I can move this to a shared crate so that webdriver_handler can also access the hashmap.
  
I think this will be enough to get tests for article passing, however, more work will need to be done to expand the HTML_ROLE_MAPPINGS to include various input types and checks that the role: 
- isn't outside required contexts (e.g. orphaned listitems)
- if synonymous, maps to the correct role (e.g. img == image, presentation == none, etc)
- doesn't override host language restrictions

I can potentially do that later, but initially I'd like to focus on the first part of this.

**Comment by jdm** [comment]
Sounds reasonable!

**Comment by alice** [comment]
I think I'd prefer if we can query the accessibility tree itself, since that will represent what the browser is actually exposing to users. I don't know that it's very helpful to pass a WPT test if that misrepresents what's actually implemented, and it'll be good to see tests start passing as we keep fleshing out the implementation.

This will be a little tricky as `webdriver_handler` is in script, while the accessibility tree is in layout. I think the most straightforward path might be to add a mechanism to retrieve the [`accesskit::Node`](https://docs.rs/accesskit/0.22.0/accesskit/struct.Node.html) for a given `Node`, and then helper methods as necessary to query that from the script side.

That's manageable enough on the accessibility tree side: [`node_for_dom_node()`](https://github.com/servo/servo/blob/fc8f70358aacf5fe20af9f10a0ecc225ca0d047b/components/layout/accessibility_tree.rs#L183) lets you get our wrapper object, [`AccessibilityNode`](https://github.com/servo/servo/blob/fc8f70358aacf5fe20af9f10a0ecc225ca0d047b/components/layout/accessibility_tree.rs#L31), which contains the latest computed properties in the internal `accesskit::Node`.

What the best strategy is for retrieving that from script might be I'm less sure. 

- First, we need to make sure the pref is enabled (I assume there's a way to enable prefs per WPT directory?) and that accesibility is active (via [`Layout::set_accessibility_active()`](https://github.com/servo/servo/blob/fc8f70358aacf5fe20af9f10a0ecc225ca0d047b/components/shared/layout/lib.rs#L406).
- The tree is only guaranteed to be up to date at the end of reflow, so it might be sensible to do something like what we do in the [`forceAccessibilityUpdate()`](https://github.com/servo/servo/blob/fc8f70358aacf5fe20af9f10a0ecc225ca0d047b/components/script/dom/testing/servotestutils.rs#L79) test util API. 
- Then, we need helper methods probably in a couple of places to retrieve the data; we could add a method on [`Layout`](https://github.com/servo/servo/blob/fc8f70358aacf5fe20af9f10a0ecc225ca0d047b/components/shared/layout/lib.rs#L270) which takes an `OpaqueNode` (or even a `ServoLayoutNode`) and returns an `accesskit::Node`, and implement that on `LayoutImpl` by calling into a helper method on `AccessibilityTree`. Finally, we could add an accessor on `Element` which calls into `Layout`.

**Comment by lolaodelola** [comment]
Thanks @alice, this makes sense! I thought I wouldn't be able to get accesskit nodes from `webdriver_handler`. 

>What the best strategy is for retrieving that from script might be I'm less sure.

Would it make sense to put the shared logic in `layout_api`? I can see comments in the code where logic that's shared between script and layout is in `layout_api`, e.g. `ScriptThreadFactory` has a comment that reads:
>This trait is part of `layout_api` because it depends on both `script_traits`
>and also `LayoutFactory` from this crate. If it was in `script_traits` there would be a
>circular dependency.

Or should I consider creating something new?

**Comment by alice** [comment]
I don't think I quite understand the question/proposal, sorry.

I worded my suggestion pretty badly; my thinking was that all the accessibility tree computation logic should continue to live in `accessibility_tree`, but that we should thread accessors through to script, something like:

- Add an `accesskit_node_for_dom_node()` function on `AccessibilityTree` which takes a `ServoLayoutNode` and returns an `Option<accesskit::Node>`
- Add a `query_accesskit_node()` function on [`Layout`](https://github.com/servo/servo/blob/fc8f70358aacf5fe20af9f10a0ecc225ca0d047b/components/shared/layout/lib.rs#L270), implemented on [`LayoutThread`](https://github.com/servo/servo/blob/d4796255ccd6b32100f3e027322378ad4def547e/components/layout/layout_impl.rs#L124) (same as we do for [`query_resolved_style()`](https://github.com/servo/servo/blob/d4796255ccd6b32100f3e027322378ad4def547e/components/layout/layout_impl.rs#L491) et al) which takes a `TrustedNodeAddress` and returns an `Option<accesskit::node>`
- Add a `computed_accesskit_node_query()` method on `Window` which works similarly to [`resolved_style_query()`](https://github.com/servo/servo/blob/d4796255ccd6b32100f3e027322378ad4def547e/components/script/dom/window.rs#L3017) (I only just thought to look at what `query_resolved_style()` etc actually do in practice, so I hadn't thought to add a step on `Window` above) but returns a (you guessed it) `Option<accesskit::Node>`
- Add a `computed_role()` method on `Element` which uses `Window::query_accesskit_node()` to retrieve all the accessibility properties and then retrieve the role value and convert it to a string - analogous to `CSSStyleDeclaration::get_computed_style()` using `resolved_style_query()`
- Use the `computed_role()` method on `Element` to implement the webdriver method.

In addition to all of that, you'll also need to enable the pref and ensure accessibility is active so the accessibility tree actually exists, and also make sure that the tree is up to date by calling `Window::layout_reflow()` immediately before querying it (again, like `resolved_style_query()` does). That in turn will probably entail adding a new value to [`QueryMsg`](https://github.com/servo/servo/blob/d4796255ccd6b32100f3e027322378ad4def547e/components/shared/layout/lib.rs#L535).

**Comment by lolaodelola** [comment]
Great, thanks @alice!

**Comment by alice** [comment]
I left a couple of key details out of my sketch above.

Firstly, we haven't accounted for the [pref](https://github.com/servo/servo/blob/7bc6f395a15494df55d0861d423fa8eb0c14614b/components/config/prefs.rs#L364). I *think* there is a way to set prefs for specific WPTs, so I'd suggest doing something like:

- In `webdriver_handlers::handle_get_computed_role()`, check the pref, and send an error response if the pref is not set.
- Make sure the pref is set for any WPTs which use the `get_computed_role` method.
  - I think you can do this by adding a `prefs` entry to (for example) [`tests/wpt/meta/webdriver/tests/classic/get_computed_role/get.py.ini`](https://github.com/servo/servo/blob/main/tests/wpt/meta/webdriver/tests/classic/get_computed_role/get.py.ini), like the one in [`tests/wpt/meta/dom/__dir__.ini`](https://github.com/servo/servo/blob/main/tests/wpt/meta/dom/__dir__.ini)

You'll also need to call [`set_needs_accessibility_update()`](https://github.com/servo/servo/blob/main/components/layout/layout_impl.rs#L769) before calling `reflow()`. It'll probably easiest just to call that method directly in your `accesskit_node_query()` method on `Window`, before calling `layout_reflow()`, i.e. something like

```rs
self.layout().set_needs_accessibility_update();
self.layout_reflow(QueryMsg::AccessKitNodeQuery);

// remaining code to fetch the accesskit node
```

**Comment by jdm** [comment]
Yes, those suggestions should work!

**Comment by lolaodelola** [comment]
- [ ] Implement role computation

**Comment by lolaodelola** [comment]
@alice @keithamus just some thoughts about implementing the next part of this work which is computing each role for each element and usecase.

**TLDR; is there a preferred way to implement the HTML-AAM spec in Servo?**

Keith and I spoke this morning and toyed with the idea of implementing traits for each element that defines `get_computed_role` to allow for the nuance and complexity for computing some of the roles. I've also had a dig around the firefox codebase and they're using a combination of a map and classes to compute the roles. In Firefox:
1. [HTMLMarkUpMap](https://searchfox.org/firefox-main/source/accessible/base/HTMLMarkupMap.inc) returns a role or an object for each element e.g. for anchor, if the tag doesn't have `href` or click listener it returns a `HyperTextAccessible`, if it does have either it returns a `HTMLLinkAccessible`
```cpp
MARKUPMAP(
    a,
    [](Element* aElement, LocalAccessible* aContext) -> already_AddRefed<LocalAccessible> {
      // An anchor element without an href attribute and without a click
      // listener should be a generic.
      if (!aElement->HasAttr(nsGkAtoms::href) &&
          !nsCoreUtils::HasClickListener(aElement)) {
        return MakeAndAddRef<HyperTextAccessible>(aElement, aContext->Document());
      }
    ...
      return MakeAndAddRef<HTMLLinkAccessible>(aElement, aContext->Document());
    },
    0)
```
2. [`HTMLLinkAccessible`](https://searchfox.org/firefox-main/source/accessible/html/HTMLLinkAccessible.cpp) defines the default role for an accessible link, as well as states, actions, etc.
```cpp
role HTMLLinkAccessible::NativeRole() const {
  if (mContent->IsMathMLElement(nsGkAtoms::a)) {
    return roles::MATHML_LINK;
  }
  return roles::LINK;
}
```
& similarly `HyperTextAccessible` defines the default role (i.e. text container which is equiv to generic), states, selectable etc.

These two files handle what's defined in HTML-AAM for the most part, we'd still need to check when to use ARIA roles and when to use HTML-AAM defined roles, and implement the role mappings for each platform API to be functional for assistive technologies.
